packages/data-schema/src/SchemaProcessor.ts

```diff
@@ -1316,6 +1419,75 @@ const mergeCustomTypeAuthRules = (
   }
 };
 
+function generateInputTypes(
+  implicitTypes: [string, any][],
+  generateInputType: boolean,
+  getRefType: ReturnType<typeof getRefTypeForSchema>,
+  authRules?: Authorization<any, any, any>[],
+  isInlineType = false,
+  definedTypes: Set<string> = new Set(),
+): string[] {
+  const generatedTypes = new Set<string>();
+
+  implicitTypes.forEach(([typeName, typeDef]) => {
+    if (!definedTypes.has(typeName)) {
+      if (isCustomType(typeDef)) {
+        const { gqlFields, implicitTypes: nestedTypes } = processFields(
+          typeName,
+          typeDef.data.fields,
+          {},
+          {},
+          getRefType,
+          undefined,
+          undefined,
+          {},
+          'dynamodb',
+          generateInputType,
+        );
+        const authString =
+          !isInlineType && authRules
+            ? mapToNativeAppSyncAuthDirectives(authRules, false).authString
+            : '';
+        const typeKeyword = generateInputType ? 'input' : 'type';
+        const customType = `${typeKeyword} ${typeName}${authString ? ` ${authString}` : ''}\n{\n  ${gqlFields.join('\n  ')}\n}`;
+        generatedTypes.add(customType);
+        definedTypes.add(typeName);
+
+        // Process nested types
+        if (nestedTypes.length > 0) {
+          const nestedGeneratedTypes = generateInputTypes(
+            nestedTypes,
+            generateInputType,
+            getRefType,
+            authRules,
+            false,
+            definedTypes,
+          );
+          nestedGeneratedTypes.forEach((type) => {
+            generatedTypes.add(type);
+          });
+        }
+      } else if (typeDef.type === 'enum') {
+        if (!definedTypes.has(typeName)) {
+          generatedTypes.add(
+            `enum ${typeName} {\n  ${typeDef.values.join('\n  ')}\n}`,
+          );
+          definedTypes.add(typeName);
+        }
+      } else if (typeDef?.data?.type === 'ref') {
+        getRefType(typeDef.data.link, typeName);
+      } else if (typeDef.type === 'scalar') {
+        generatedTypes.add(`scalar ${typeName}`);
+        definedTypes.add(typeName);
+      } else {
+        console.warn(`Unexpected type definition for ${typeName}:`, typeDef);
```

Comment:
Is this useful?  I don't see it being tested. What does this else mean in this case? If it doesn't matter to the end user, lets remove it. If its a catastrophic logic error, maybe we should raise. If there's logic for it, we should have a unit test for it, since its part of the user experience.